test/e2e/main.go

```diff
@@ -22,21 +22,154 @@ import (
+		return
+	}
+	return
+}
+
+func checkCreatedJobs(kubeCli kubernetes.Interface, name string, original, created *spec.TfJob) error {
+	// Loop over each replica and make sure the expected resources were created.
+	for _, r := range original.Spec.ReplicaSpecs {
+		baseName := strings.ToLower(string(r.TfReplicaType))
+
+		for i := 0; i < int(*r.Replicas); i += 1 {
+			jobName := fmt.Sprintf("%v-%v-%v-%v", fmt.Sprintf("%.40s", original.Metadata.Name), baseName, created.Spec.RuntimeId, i)
+
+			_, err := kubeCli.BatchV1().Jobs(Namespace).Get(jobName, metav1.GetOptions{})
+
+			if err != nil {
+				return fmt.Errorf("the Tfob %v did not create Job %v for ReplicaType %v Index %v", name, jobName, r.TfReplicaType, i)
+			}
+		}
+	}
+	return nil
+}
+
+func getJobNames(original, created *spec.TfJob) (jobs map[string]bool) {
+	jobs = make(map[string]bool)
+	// Loop over each replica and make sure the expected resources are being deleted.
+	for _, r := range original.Spec.ReplicaSpecs {
+		baseName := strings.ToLower(string(r.TfReplicaType))
+
+		for i := 0; i < int(*r.Replicas); i += 1 {
+			jobName := fmt.Sprintf("%v-%v-%v-%v", fmt.Sprintf("%.40s", original.Metadata.Name), baseName, created.Spec.RuntimeId, i)
+			jobs[jobName] = true
+		}
+	}
+	return
+}
+
+func updateJobNames(kubeCli kubernetes.Interface, jobs map[string]bool) {
+	for k := range jobs {
+		_, err := kubeCli.BatchV1().Jobs(Namespace).Get(k, metav1.GetOptions{})
+		if k8sErrors.IsNotFound(err) {
+			// Deleting map entry during loop is safe.
+			// See: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-golang-map-within-a-range-loop
+			delete(jobs, k)
+		} else {
+			log.Infof("Job %v still exists", k)
+		}
+	}
+}
+
+func waitForDeletion(kubeCli kubernetes.Interface, name, tbDeployName string, original, created *spec.TfJob) (err error) {
+	// Define sets to keep track of Job controllers corresponding to Replicas
+	// that still exist.
+	jobs := getJobNames(original, created)
+	isTBDeployDeleted := false
+
+	// Wait for all jobs and deployment to be deleted.
+	for endTime := time.Now().Add(*timeout); time.Now().Before(endTime) && (len(jobs) > 0 || !isTBDeployDeleted); {
+		updateJobNames(kubeCli, jobs)
+
+		if !isTBDeployDeleted {
+			// Check that TensorBoard deployment is being deleted
+			_, err = kubeCli.ExtensionsV1beta1().Deployments(Namespace).Get(tbDeployName, metav1.GetOptions{})
+			if k8sErrors.IsNotFound(err) {
+				isTBDeployDeleted = true
+			} else {
+				log.Infof("TensorBoard deployment %v still exists for TfJob %v", tbDeployName, name)
+			}
+		}
+
+		if len(jobs) > 0 || !isTBDeployDeleted {
+			time.Sleep(5 * time.Second)
+		}
+	}
+
+	if len(jobs) > 0 {
+		err = fmt.Errorf("Not all Job controllers were successfully deleted for TfJob %v.", name)
+		return
+	}
+
+	if !isTBDeployDeleted {
+		err = fmt.Errorf("TensorBoard deployment %v was not successfully deleted for TfJob %v.", tbDeployName, name)
+		return
```

Comment:
remove this return

Reply:
this is early return, prefer to keep it future-proof